### PR TITLE
Various minor trait improvements

### PR DESCRIPTION
**Pull request by flodiebold**
- lower bounds on trait definition, i.e. super traits
 - use super traits for associated types
 - use traits from where clauses and their super traits for method resolution
 - lower fn-like paths (i.e. `Fn(X, Y) -> Z`)
 - pass the environment to Chalk in the correct way to make elaboration work, i.e. inferring things like `T: Clone` from `T: Copy`. The clauses need to be wrapped in `FromEnv` clauses for that to work, which I didn't do before.
 - add some tests for closure inference already

### CODE REVIEW

File: crates/ra_hir/src/code_model.rs [hunk 1]
```diff
@@ -829,7 +868,18 @@ impl Trait {
...
+
+    pub fn associated_type_by_name_including_super_traits(
+        self,
+        db: &impl HirDatabase,
+        name: &Name,
+    ) -> Option<TypeAlias> {
+        iter::once(self)
+            .chain(self.all_super_traits(db))
+            .unique()
+            .find_map(|t| t.associated_type_by_name(db, name))
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs]: It seems like `unique` before `.find_map` is a no-op?

File: crates/ra_hir/src/code_model.rs [hunk 2]
```diff
@@ -849,20 +846,23 @@ impl Trait {
...
+            // yeah this is quadratic, but trait hierarchies should be flat
+            // enough that this doesn't matter
+            for tt in t.direct_super_traits(db) {
+                if !result.contains(&tt) {
+                    result.push(tt);
+                }
+            }
+            i += 1;
+        }
+        result
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs]: hehe, I remember writing this kind of code quite a few times already (though some of thouse might be in IntelliJ?). Perhaps we should create some kind of `.explore()` iterator adaptor one day

File: crates/ra_hir/src/code_model.rs [hunk 3]
```diff
@@ -849,20 +846,23 @@ impl Trait {
...
-    }
-
-    fn all_super_traits_inner<'a>(
-        self,
-        db: &'a impl HirDatabase,
-    ) -> impl Iterator<Item = Trait> + 'a {
-        iter::once(self).chain(self.direct_super_traits(db).into_iter().flat_map(move |t| {
-            iter::once(t).chain(Box::new(t.all_super_traits(db)) as Box<dyn Iterator<Item = Trait>>)
-        }))
+    pub fn all_super_traits(self, db: &impl HirDatabase) -> Vec<Trait> {
```
Comment (matklad) [inline crates/ra_hir/src/code_model.rs]: it seems like it can return an iterator, though it probably does not makes sense to do.

Comment (flodiebold) [reply crates/ra_hir/src/code_model.rs]: I tried that at first, but that implementation didn't deal correctly with cycles... I think doing it with iterator adapters and handling cycles would be quite complicated, and I'm pretty sure we'd have to keep the set of already visited traits in any case, so it probably wouldn't help :thinking:

**Pull request merged**